librz/bin/p/bin_nin3ds.c

```diff
@@ -7,136 +8,329 @@
+			continue;
 		}
+		rz_list_append(ret, sect);
 	}
 
-	/* Append sections or free them if file is corrupt to avoid memory leaks */
-	for (i = 0; i < 4; i++) {
-		if (sections[i]) {
-			if (corrupt) {
-				free(sections[i]);
-			} else {
-				rz_list_append(ret, sections[i]);
-			}
-		}
+	return ret;
+}
+
+static RzList /*<RzBinAddr *>*/ *n3ds_entries(RzBinFile *bf) {
+	N3DSFirmHdr *hdr = NULL;
+	RzList *ret = NULL;
+	RzBinAddr *ptr9 = NULL, *ptr11 = NULL;
+
+	if (!bf || !bf->buf) {
+		return NULL;
 	}
-	if (corrupt) {
+
+	hdr = rz_bin_file_get_hdr(bf);
+	ret = rz_list_newf(free);
+
+	if (!ret ||
+		!(ptr9 = RZ_NEW0(RzBinAddr)) ||
+		!(ptr11 = RZ_NEW0(RzBinAddr))) {
 		rz_list_free(ret);
+		free(ptr9);
 		return NULL;
 	}
 
-	return ret;
-}
+	/* ARM9 entry point */
+	ptr9->vaddr = hdr->arm9_ep;
+	rz_list_append(ret, ptr9);
 
-static RzList /*<RzBinAddr *>*/ *entries(RzBinFile *bf) {
-	RzList *ret = rz_list_new();
-	RzBinAddr *ptr9 = NULL, *ptr11 = NULL;
+	/* ARM11 entry point */
+	ptr11->vaddr = hdr->arm11_ep;
+	rz_list_append(ret, ptr11);
 
-	if (bf && bf->buf) {
-		if (!ret) {
-			return NULL;
-		}
-		ret->free = free;
-		if (!(ptr9 = RZ_NEW0(RzBinAddr))) {
-			rz_list_free(ret);
-			return NULL;
+	for (size_t i = 0; i < 4; i++) {
+		N3DSFirmSectHdr *shdr = &hdr->sections[i];
+		if (!shdr->size) {
+			continue;
 		}
-		if (!(ptr11 = RZ_NEW0(RzBinAddr))) {
-			rz_list_free(ret);
-			free(ptr9);
-			return NULL;
-		}
-
-		/* ARM9 entry point */
-		ptr9->vaddr = loaded_header.arm9_ep;
-		rz_list_append(ret, ptr9);
 
-		/* ARM11 entry point */
-		ptr11->vaddr = loaded_header.arm11_ep;
-		rz_list_append(ret, ptr11);
+		ut64 beg_section = shdr->address;
+		ut64 end_section = shdr->address + shdr->size;
+		if (beg_section <= ptr9->vaddr && ptr9->vaddr < end_section) {
+			ut32 diff = shdr->address - hdr->arm9_ep;
+			ptr9->paddr = shdr->offset + diff;
+		} else if (beg_section <= ptr11->vaddr && ptr11->vaddr < end_section) {
+			ut32 diff = shdr->address - hdr->arm11_ep;
+			ptr11->paddr = shdr->offset + diff;
+		}
 	}
+
 	return ret;
 }
 
-static RzBinInfo *info(RzBinFile *bf) {
+static RzBinInfo *n3ds_info(RzBinFile *bf) {
 	RzBinInfo *ret = RZ_NEW0(RzBinInfo);
 	if (!ret) {
 		return NULL;
 	}
 
-	if (!bf || !bf->buf) {
-		free(ret);
-		return NULL;
-	}
-
 	ret->type = strdup("FIRM");
 	ret->machine = strdup("Nintendo 3DS");
 	ret->os = strdup("n3ds");
 	ret->arch = strdup("arm");
 	ret->has_va = true;
 	ret->bits = 32;
-
 	return ret;
 }
 
+static RzBinFileHash *n3ds_hash_buffer(const char *name, const ut8 *hash, size_t size) {
+	RzBinFileHash *fh = RZ_NEW0(RzBinFileHash);
+	if (!fh) {
+		return NULL;
+	}
+	fh->type = strdup(name);
+	fh->hex = rz_hex_bin2strdup(hash, size);
+	return fh;
+}
+
+static RzList /*<RzBinFileHash *>*/ *n3ds_hashes(RzBinFile *bf) {
+	if (!bf || !bf->buf) {
+		return NULL;
+	}
+
+	N3DSFirmHdr *hdr = rz_bin_file_get_hdr(bf);
+	RzBinFileHash *fh = NULL;
+	RzList *list = rz_list_newf((RzListFree)rz_bin_file_hash_free);
```

Comment:
You can move creation of the new list and erro checking before defining any of those variables, will allow exit even earlier if there is an error